catalog/cataloger_merge.go

```diff
@@ -52,49 +54,46 @@ func (c *cataloger) Merge(ctx context.Context, repository, leftBranch, rightBran
 				Limit: -1,
 			},
 		}
-
-		scanner, err := NewDiffScanner(tx, params)
+		relation, err := getRefsRelationType(tx, params)
 		if err != nil {
 			return nil, err
 		}
-		if scanner.Relation == RelationTypeSame {
+		if relation == RelationTypeSame {
 			return nil, ErrSameBranchMergeNotSupported
 		}
 		nextCommitID, err := getNextCommitID(tx)
 		if err != nil {
 			return nil, err
 		}
-		// do the merge based on the relation
 		previousMaxCommitID, err := getLastCommitIDByBranchID(tx, rightID)
 		if err != nil {
 			return nil, err
 		}
-
-		differences := make(mergeBatchRecords, 0, MergeBatchSize)
+		mergeBatchChan := make(chan mergeBatchRecords, MergeBatchChanBuffer)
+		var workerExitFlag bool
```

Comment:
using variables without locking to communicate state between goroutine is a race issue

Reply:
This is an exit flag - it means that one goroutine will read the value and one will update the value.
Any unsynchronized access to shared variables is a race.